distributed/tests/test_client.py

```diff
@@ -4720,14 +4714,18 @@ class Foo:
         def __getstate__(self):
             raise MyException()
 
-    with pytest.raises(MyException):
+    # Notice, because serialization is delayed until `distributed.batched`
+    # we get a `CancelledError` exception.
+    # Before <https://github.com/dask/distributed/pull/4699> the serialization
+    # happed immediately in `submit()`, which would raise the `MyException`.
+    with pytest.raises(CancelledError):
         future = c.submit(identity, Foo())
 
-    futures = c.map(inc, range(10))
-    results = await c.gather(futures)
+        futures = c.map(inc, range(10))
+        results = await c.gather(futures)
 
-    assert results == list(map(inc, range(10)))
-    assert a.data and b.data
+        assert results == list(map(inc, range(10)))
+        assert a.data and b.data
```

Comment:
Hrm, you mentioned this in meeting a couple of weeks ago.  I see now how this is unfortunate.  

I would expect this test to now be written as 

```python
with pytest.raises(CancelledError):
    await c.submit(identity, Foo())
```

I wouldn't expect the other lines here to be indented.  In general seeing assert statements under a `raises` context manager is a sign that something is unclean :)

Reply:
I have changed it to make it more clear what is going on:
```python
    # Notice, because serialization is delayed until `distributed.batched`
    # we don't get an exception immediately. The exception is raised and logged
    # when the ongoing communication between the client the scheduler encounters
    # the `Foo` class. Before <https://github.com/dask/distributed/pull/4699>
    # the serialization happened immediately in `submit()`, which would raise the
    # `MyException`.
    with captured_logger("distributed") as caplog:
        future = c.submit(identity, Foo())
        await asyncio.sleep(c.scheduler_comm.interval)
    # Check that the serialization error was logged
    assert "Failed to serialize" in caplog.getvalue()
```
I'm curious, what do you think of the approach? We cannot easily catch the exception because it happens as part of the ongoing communication and not in the `submit()` call, but at least we log the exception.